### DISCUSSION TITLE
Script is not saved when SDCard is enabled

### DISCUSSION THREAD

**Discussion by SZenglein** [arendst/Tasmota#19049] [Q&A]
### PROBLEM DESCRIPTION
If an SD card is enabled and connected, script.txt can not be saved or loaded.

I would have liked to do some smart meter reading locally (without external broker) using the smart meter interface, berry and an SD card. I have connected the SD card via SDIO (SD_MMC).

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): esp32-devkitc v4 wrover-b 8MB
- [x] Tasmota binary firmware version number used: 12.5.0.4(tasmota32)
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: platformio cli
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
18:57:40.495 CMD: Backlog Template; Module; GPIO 255
18:57:40.496 SRC: WebConsole from 192.168.48.190
18:57:40.498 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 26, Pld -99, Data 'Template; Module; GPIO 255'
18:57:40.519 SRC: Backlog
18:57:40.521 CMD: Grp 0, Cmd 'TEMPLATE', Idx 1, Len 0, Pld -99, Data ''
18:57:40.526 RSL: RESULT = {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
18:57:40.744 SRC: Backlog
18:57:40.746 CMD: Grp 0, Cmd 'MODULE', Idx 1, Len 0, Pld -99, Data ''
18:57:40.748 RSL: RESULT = {"Module":{"1":"ESP32-DevKit"}}
18:57:40.994 SRC: Backlog
18:57:40.996 CMD: Grp 0, Cmd 'GPIO', Idx 1, Len 3, Pld 255, Data '255'
18:57:41.007 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"8864":"SDIO D0"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"8832":"SDIO CLK"},"GPIO15":{"8800":"SDIO CMD"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}
18:57:46.664 WIF: Prüfe Verbindung...


```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua

```
- [x] Provide the output of this command: `Status 0`:
```lua
18:58:35.792 SRC: WebConsole from 192.168.48.190
18:58:35.794 CMD: Grp 0, Cmd 'STATUS', Idx 1, Len 1, Pld 0, Data '0'
18:58:35.797 RSL: STATUS = {"Status":{"Module":1,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_CD1970","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
18:58:35.807 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32-DE.bin","RestartReason":"Software reset CPU","Uptime":"0T00:17:53","StartupUTC":"2023-06-26T17:40:42","Sleep":50,"CfgHolder":4617,"BootCount":35,"BCResetTime":"2023-06-24T20:21:35","SaveCount":63}}
18:58:35.815 RSL: STATUS2 = {"StatusFWR":{"Version":"12.5.0.4(tasmota32)","BuildDateTime":"2023.06.25 20:27:23","Core":"2_0_10","SDK":"4.4.5.230614","CpuFrequency":160,"Hardware":"ESP32-D0WD-V3 rev.3","CR":"405/699"}}
18:58:35.818 RSL: STATUS3 = {"StatusLOG":{"SerialLog":4,"WebLog":4,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["sess",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
18:58:35.834 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":1845,"Free":1034,"Heap":124,"StackLowMark":3,"PsrMax":4093,"PsrFree":4049,"ProgramFlashSize":4096,"FlashSize":8192,"FlashChipId":"174068","FlashFrequency":40,"FlashMode":"DIO","Features":["00000407","979AD7CF","001DA001","B7F7BFCF","05DA9BC0","E0360DCF","480840D2","20200000","D4BC482D","000A80B1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,66,67,68,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,31,34,37,39,40,42,43,45,51,52,53,55,56,58,59,64,66,67,74,85,92,95,98,103,105,127","I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,65,69,76,77"}}
18:58:35.846 BRY: GC from 8410 to 4876 bytes, objects freed 9/46 (in 1 ms) - slots from 52/122 to 51/122
18:58:35.851 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-CD1970-6512","IPAddress":"192.168.48.30","Gateway":"192.168.48.161","Subnetmask":"255.255.255.0","DNSServer1":"192.168.48.161","DNSServer2":"0.0.0.0","Mac":"78:E3:6D:CD:19:70","IP6Global":"","IP6Local":"fe80::7ae3:6dff:fecd:1970","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"192.168.48.161","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00","IP6Global":"","IP6Local":""},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
18:58:35.862 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_CD1970","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:58:35.868 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-06-26T17:58:35","Local":"2023-06-26T18:58:35","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"04:48","Sunset":"20:57"}}
18:58:35.877 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-06-26T18:58:35"}}
18:58:35.881 BRY: GC from 5879 to 4183 bytes, objects freed 4/46 (in 1 ms) - slots from 51/122 to 51/122
18:58:35.887 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-06-26T18:58:35","Uptime":"0T00:17:53","UptimeSec":1073,"Heap":124,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":46},"Wifi":{"AP":1,"SSId":"sess","BSSId":"A6:74:7D:D7:DD:14","Channel":11,"Mode":"11n","RSSI":100,"Signal":-21,"LinkCount":1,"Downtime":"0T00:00:03"}}}
18:58:47.196 WIF: Prüfe Verbindung...

```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua



```

### TO REPRODUCE
_Steps to reproduce the behavior:_
If the sdcard is disconnected, script.txt is created normally and run on startup.
If the sdcard is connected, the script.txt is attempted to be written to SD (but fails), leaving an empty file. It also does not work if script.txt exists on the flash while the sdcard is connected.

This is the configuration I used:
user_config_override.h
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

#define USE_UFILESYS
#define USE_SDCARD
```

### EXPECTED BEHAVIOUR
Can save script.txt when sd is enabled.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
Well, then you should do something about 
> If the sdcard is connected, the script.txt is attempted to be written to SD (but fails)

You told no details about what "fails" means.

Until you have a working file system, do not expect normal operation around using scripts. It is a feature to have scripts on the file system, not a bug.

**Comment by SZenglein** [comment]
The file system is working. I can create and edit files.

I don't know why it left an empty file, but after trying a few times it created the file and actually ran the script afterwards.

So at this point I see two issues:
* script.txt saving to SD is inconsistent
* is script.txt suppised to be saved on SD card? I would expect the default location to not change from / to /sd/

**Comment by sfromis** [comment]
Well, what I meant was that whatever you meant by "but fails" looks like something in need of understanding the details of .....

**Comment by Jason2866** [comment]
Please describe exactly what is not working. Writing to SD Card is the wanted behaviour.
No one can help just with the statement "it fails"

**Comment by DiegoB80** [comment]
Dears, 
I think to have the same problem.
I have a script saved on FlashFS. It works.
When I insert the SDCard and I restart my ESP32S3, the system seems to read by default from SDCard and tasmota doesn't see the script (the script.txt file is saved on internal Flash).
When I remove SDCard and I restart the ESP32S3 the script is correctly read (but without flag "Script enable").
In my opinion the file script.txt should be read always from FlashFS independently if I inserted a SDCard.
The same think happens with Berry.
I have an image saved on FlashFS (Inverter.png).
With this command I load the image: 
imgInverter.set_src("A:/Inverter.png")
if I insert a SDCard and I restart my tasmota, the image isn't loaded. I remove the SDCard, I restart my Tasmota and the image is loaded.

I think that the SDCard become the default/primary drive for tasmota. 

Diego

**Reply by sfromis** [reply]
Since you mention Berry, my experience is that when booting, and having an `autoexec.be` (or `autoexec.bat`) on both internal flash and external SD card, the one from flash will be executed. Also getting the flash file system with a test case like `import path print(path.listdir('/'))`, only getting the SD card files if I instead use `path.listdir('/sd/')`.

I've not tested with scripting.
